service/history/workflow/cache/cache.go

```diff
@@ -178,32 +191,36 @@ func (c *CacheImpl) GetOrCreateWorkflowExecution(
 
 func (c *CacheImpl) getOrCreateWorkflowExecutionInternal(
 	ctx context.Context,
+	shardContext shard.Context,
 	namespaceID namespace.ID,
 	execution commonpb.WorkflowExecution,
 	handler metrics.Handler,
 	forceClearContext bool,
 	lockPriority workflow.LockPriority,
 ) (workflow.Context, ReleaseCacheFunc, error) {
 
-	key := definition.NewWorkflowKey(namespaceID.String(), execution.GetWorkflowId(), execution.GetRunId())
-	workflowCtx, cacheHit := c.Get(key).(workflow.Context)
+	cacheKey := Key{
+		WorkflowKey: definition.NewWorkflowKey(namespaceID.String(), execution.GetWorkflowId(), execution.GetRunId()),
+		ShardUUID:   shardContext.GetOwner(),
+	}
+	workflowCtx, cacheHit := c.Get(cacheKey).(workflow.Context)
 	if !cacheHit {
 		handler.Counter(metrics.CacheMissCounter.GetMetricName()).Record(1)
 		// Let's create the workflow execution workflowCtx
-		workflowCtx = workflow.NewContext(c.shard, key, c.logger)
-		elem, err := c.PutIfNotExist(key, workflowCtx)
+		wfCtxLogger := log.With(c.logger, tag.ShardID(shardContext.GetShardID()))
```

Comment:
use shardContext.GetLogger()?

In the original comment I actually mean the to tag the usage of c.logger on L288.

Reply:
Updated both and not DI the logger and metrics handler into cache.